bundles/org.openhab.ui/web/src/assets/definitions/blockly/utils.js

```diff
@@ -13,31 +13,128 @@ export function addOSGiService (serviceName, serviceClass) {
 }
 
 /*
 * Add ZoneDateTime and DateTimeFormatter support to rule
+* Supports a number of common formats out of the box like
+* - 2022-04-10 (10)
+* - 2022-04-09 10:11 (16)
+* - 2022-04-09 10:11:12 (19)
+* - 2022-04-10T18:50:55+02:00 (25)
+* - 2022-04-10T18:50:55.537 (23)
+* - 2022-04-10T18:50:55.537123 (26)
+* - 2022-04-10T18:50:55.537+02:00 (29)
+* - 2022-04-10T18:50:55.537123+02:00 (32) -> local date time standard output
+* - 2022-05-01T12:28:50.662+0200 (28) -> OH standard output format
+* - all of the above work with or without the "T"
 */
 export function addDateSupport () {
-  const dtf = Blockly.JavaScript.provideFunction_(
+  let dtf = Blockly.JavaScript.provideFunction_(
     'dtf',
     ['var ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + ' = Java.type("java.time.format.DateTimeFormatter");'])
-  const zdt = Blockly.JavaScript.provideFunction_(
+
+  let zdt = Blockly.JavaScript.provideFunction_(
     'zdt',
     ['var ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + ' = Java.type("java.time.ZonedDateTime");'])
-  const getZonedDateTime = Blockly.JavaScript.provideFunction_(
-    'getZonedDateTime',
-    [
-      'function ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + ' (datetime) {',
-      '  switch (datetime.length) {\n' +
-      `    case 10: return ${zdt}.parse(datetime + ' 00:00:00 +00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ss z'));\n` +
-      `    case 19: return ${zdt}.parse(datetime + ' +00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ss z'));\n` +
-      `    case 26: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ss z'));\n` +
+
+  let getzdt = Blockly.JavaScript.provideFunction_(
+    'getZonedDateTime', [
+      '/* Try to detect the format based on its length */',
+      'function ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + '(datetime) {',
+      '  datetime = datetime.replace(\'T\', \' \')',
+      '  switch (datetime.length) {',
+      `    case 10: return ${zdt}.parse(datetime + ' 00:00:00+00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 16: return ${zdt}.parse(datetime + ':00+00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 19: return ${zdt}.parse(datetime + '+00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 25: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 23: return ${zdt}.parse(datetime + ' +00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSS z'));`,
+      `    case 26: return ${zdt}.parse(datetime + ' +00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSSS z'));`,
+      `    case 29: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSz'));`,
+      `    case 32: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSSSSz'));`,
+      `    case 28: return ${zdt}.parse(datetime.slice(0,26) + ':' + datetime.slice(26,28), dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSz'));`,
+      `    default: return ${zdt}.parse(datetime);`,
+      '  }',
+      '}'
```

Comment:
Honestly (and I missed it the previous time), I'm not the biggest fan of this implementation, especially as you have a block that allows the user to submit arbitrary strings to this function. Moreover it's kind of opaque when you see this big switch/case block added to your script. Surely extracting values with a all-encompassing regular expression then using [`ZonedDateTime.of`](https://howtodoinjava.com/java/date-time/zoneddatetime-class/#22-create-zoneddatetime-with-values) would have been more elegant.
I'll accept still as-is, considering that eventually Nashorn support will be dropped and then we'll be able to use functions from openhab-js (or add support directly there) which would make these kinds of utility functions obsolete.

Reply:
Not to defend myself but to explain why I did it: The original block
<img width="270" alt="image" src="https://user-images.githubusercontent.com/5937600/170953307-09521a38-18d3-4614-a42f-6b7e2bc55248.png">
was only able to take a this simple date format and I got complaints in the forum that this block was always resetting the time. I then started to think about what would be the most common patterns that someone would use as a string (including of what is being produced by OH itself), so I added all this cases (which was a pretty tedious job and not really fun) like these (and others)

<img width="390" alt="image" src="https://user-images.githubusercontent.com/5937600/170955090-1fe91c0c-e1fd-4683-8e76-9090f1e318dd.png">

I totally agree that the code itself with the switch based on the length of the string isn't really the nicest way of doing it but I wasn't able to do it better myself. I even started doing it via regex but actually failed to do it correctly .

So, yes, I agree the code could probably be done better and hopefully it can be approved eventually.
Thanks for pointing out ZonedDateTime.of . I hadn't known that one. I never stop learning on dates and times...